### DISCUSSION TITLE
Feedback on the new C-API wrapper welcome

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#583] [General]
Thanks to @siiky we now have a new C-API, see #553. This currently only covers the reading part and is implemented as a "wrapper". If this turns out to be useful, I intend to merge that with the core library, such that is comes with every zxing-cpp installation out of the box. Before I do that I'd like to get some feedback from the community, though.

If you have a use for this, please drop me a line. If you have suggestions on how to improve the API, let me know as well!

EDIT: added a [README](https://github.com/zxing-cpp/zxing-cpp/tree/master/wrappers/c) with more info.

**Comment by hoisagen** [comment]
always love me some c api's :) it is very useful indeed

**Reply by axxel** [reply]
Anyone except @hoisagen who has tried this? There is a new release on the horizon and I'm contemplating adding this new API to the library itself. But not without some solid real-world experience proving its adequacy. @siiky @peterlinux2010?

**Comment by siiky** [comment]
Hey @axxel! The C API seems to behave well so far. I noticed a slight RSS memory increase (40~60MiB out of ~270MiB) with high resolution images (2592x1944). Unsure of the cause yet. Lower resolutions have lower memory increase too, that's why this is not worrying for me.

I had trouble compiling because of the `reinterpret_cast` and `static_cast` in `zxing_BarcodeFormatsFromString()` (now fixed locally). Also unsure why, maybe something related to `cgo`?

And for my use case I added `zxing_DecodeHints_setMaxNumberOfSymbols()` (though maybe that's unnecessary if using `zxing_ReadBarcode()`?).

The only thing I'm still doing myself is getting the result out, to avoid one unnecessary allocation. This is probably not a problem if using C directly; it's only a problem for me because getting the result into Go's GC implies another copy.

**Reply by axxel** [reply]
> Unsure of the cause yet. Lower resolutions have lower memory increase too, that's why this is not worrying for me.

Interesting. One thing that would increase your memory usage depending on the input size is if you switched from C++17 to C++20 because the DataMatrix detector behaves differently then and uses an internal cache with the size of the frame (i.e. 5MB with 2592x1944). If you find out where or with what change you got this increase, please let me know.

> I had trouble compiling because of the `reinterpret_cast` and `static_cast` in `zxing_BarcodeFormatsFromString()` (now fixed locally). Also unsure why, maybe something related to `cgo`?

Can you let me know the compiler error message and how you fixed it?

> And for my use case I added `zxing_DecodeHints_setMaxNumberOfSymbols()` (though maybe that's unnecessary if using `zxing_ReadBarcode()`?).

If you post a PR with that addition, I'd be happily merging it.

> The only thing I'm still doing myself is getting the result out, to avoid one unnecessary allocation. This is probably not a problem if using C directly; it's only a problem for me because getting the result into Go's GC implies another copy.

What do mean by "getting out the result"?

Looks like you were actually interested in a go-wrapper, not a c-wrapper? Do you think it would make sense to add go language bindings to the project directly?

**Reply by siiky** [reply]
> One thing that would increase your memory usage depending on the input size is if you switched from C++17 to C++20

The difference was between the previous version using my C hack wrapper, and the current version using the new official C API. AFAIK nothing significant changed in the build process. `cgo` doesn't explicitly set a C or C++ standard, but I can try setting it myself explicitly and report back. Thanks for the tip!

> What do mean by "getting out the result"?

I just mean using `zxing_Result_text()`/`zxing_Result_bytes()`.

> Looks like you were actually interested in a go-wrapper, not a c-wrapper? Do you think it would make sense to add go language bindings to the project directly?

For our usecase yes :) However, a Go wrapper/library sounds like _a lot_ more work than a C wrapper. Plus, I'm not very comfortable with Go as I am with C. Hacking something together to make the Go program work is one thing, but I wouldn't really know how to make a "real" Go library.

> Can you let me know the compiler error message and how you fixed it?
> (...)
> If you post a PR with that addition, I'd be happily merging it.

Sure! I can do that tomorrow.

**Reply by axxel** [reply]
> I just mean using `zxing_Result_text()`/`zxing_Result_bytes()`.

And how is your own solution different than what we have in the wrapper?

**Reply by siiky** [reply]
I access the data of the `ByteArray` in the `Result` directly, keeping the latter until after the new copy owned by Go's GC is made.

```go
	r := C.zxing_read_barcode(Cjpeg, Cjpeglen, format)
	if r == nil {
		return nil
	}
	defer C.zxing_Result_delete(r)

	if !C.zxing_Result_isValid(r) {
		return nil
	}

	bytes := C.zxing_Result_Bytes(r)
	size := C.zxing_ByteArray_size(bytes)
	data := C.zxing_ByteArray_data(bytes)
	return C.GoBytes(data, size)
```

The functions from C's side (not in the official C API):

```c
zxing_Result *          zxing_read_barcode   (const void * jpeg, unsigned int len, const char * fmt);
const zxing_ByteArray * zxing_Result_Bytes   (const zxing_Result * r);
int                     zxing_ByteArray_size (const zxing_ByteArray * ba);
const void *            zxing_ByteArray_data (const zxing_ByteArray * ba);
```

And now it occurs to me that maybe `Result.bytes()` may itself make a copy..?

**Reply by axxel** [reply]
> And now it occurs to me that maybe Result.bytes() may itself make a copy..?

No, it does not.

So you are trying to avoid one copy of at most a few hundred bytes? May I ask why? I can assure you that this copy operation is completely dwarfed by the amount of copying and general CPU utilization that is required to come to the point where you have this result.

Does your `zxing_read_barcode` function use `stbi_load...` to decode the jpeg internally? May I have a look at the implementation of that function?

**Reply by siiky** [reply]
I'm running now a comparison benchmark between `-std=c++17` and `-std=c++20`.

---

> Can you let me know the compiler error message and how you fixed it?

The error message:

```
zxing-c.cpp: In function ‘zxing_BarcodeFormats zxing_BarcodeFormatsFromString(const char*)’:
zxing-c.cpp:50:59: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   50 |                 return static_cast<zxing_BarcodeFormats>(*reinterpret_cast<BarcodeFormat*>(&format));
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The diff:

```diff
diff --git a/wrappers/c/zxing-c.cpp b/wrappers/c/zxing-c.cpp
index befb8b77..c80e16d4 100644
--- a/wrappers/c/zxing-c.cpp
+++ b/wrappers/c/zxing-c.cpp
@@ -46,8 +46,9 @@ zxing_BarcodeFormats zxing_BarcodeFormatsFromString(const char* str)
        if (!str)
                return {};
        try {
-               auto format = BarcodeFormatsFromString(str);
-               return static_cast<zxing_BarcodeFormats>(*reinterpret_cast<BarcodeFormat*>(&format));
+               BarcodeFormats formats = BarcodeFormatsFromString(str);
+               BarcodeFormat * format = reinterpret_cast<BarcodeFormat*>(&formats);
+               return static_cast<zxing_BarcodeFormats>(*format);
        } catch (...) {
                return zxing_BarcodeFormat_Invalid;
        }
```

---

> So you are trying to avoid one copy of at most a few hundred bytes? May I ask why? I can assure you that this copy operation is completely dwarfed by the amount of copying and general CPU utilization that is required to come to the point where you have this result.

I know, but I have nothing to lose here, and I already had the code anyway.

---

> Does your `zxing_read_barcode` function use `stbi_load...` to decode the jpeg internally? May I have a look at the implementation of that function?

```c
zxing_Result * zxing_read_barcode (const void * jpeg, unsigned int len, const char * fmt)
{
	zxing_Result * ret = NULL;
	zxing_ImageView * iv = NULL;

	int width = 0;
	int height = 0;
	uint8_t * buf = (uint8_t *) jpeg_decode(jpeg, len, &width, &height, JPEG_RGB);
	if (!buf) return NULL;

	zxing_DecodeHints * hints = zxing_DecodeHints_new();
	if (!hints) goto free_buf;

	zxing_DecodeHints_setEanAddOnSymbol(hints, zxing_EanAddOnSymbol_Ignore);
	zxing_DecodeHints_setFormats(hints, zxing_BarcodeFormatFromString(fmt));
	zxing_DecodeHints_setTryRotate(hints, true);
	zxing_DecodeHints_setReturnErrors(hints, false);
	zxing_DecodeHints_setMaxNumberOfSymbols(hints, 1);

	iv = zxing_ImageView_new(buf, width, height, zxing_ImageFormat_RGB, 0, 0);
	if (!iv) goto free_hints;

	ret = zxing_ReadBarcode(iv, hints);

	zxing_ImageView_delete(iv);

free_hints:
	zxing_DecodeHints_delete(hints);

free_buf:
	jpeg_free(buf);

	return ret;
}
```

**Reply by siiky** [reply]
There's no significant difference in memory usage between C++17 and C++20. At lower resolutions there's barely any difference at all (1\~2MB). At the highest resolution there's a tiny bit more noise but the difference is still very small (10\~15MB, with C++20 being more stable and generally using less).

**Reply by axxel** [reply]
Feedback:
 * `returnErrors` is set to false by default
 * `setMaxNumberOfSymbols(1)` is implicit if you use the `ReadBarcode` (as opposed to `ReadBarcodes`) function
 * If you can, you should try asking your jpeg decoder to return greyscale (or Yuv) pixels and then use `ImageFormat::Lum`, this will save you a multitude in memcopies and cpu cycles of what you save by avoiding the `Result::bytes()` copy.

Regarding the memory usage: so c++20 is not responsible for your noticed increase. Any other hint what might have caused it?

**Reply by siiky** [reply]
Hey, sorry again for the long silence.

Thanks for the grayscale tip -- I had already looked for it in `ImageFormat` but didn't realize it was `Lum`.

I don't have any more ideas as of now about the memory use increase. If I learn anything new about this I'll let you know!

**Comment by axxel** [comment]
I added a `zxing_LastErrorMsg()` function to be able to retrieve errors like invalid parameter passing to `ReadBarcodes()`, which throws an exception that was not caught earlier and would have crashed a c client application.

**Reply by axxel** [reply]
I added some more functions (getters for `ReaderOptions` and a `zxing_free()` to free the memory allocated internally for all the `char *` stuff. I also included the wrapper code directly into the main library (when `BUILD_C_API=ON` is configured). But this is still not released, so the API may still without notice.

**Comment by axxel** [comment]
While working on the Rust wrapper and now looking into a .NET wrapper, I realized that having `zxing_ImageView` and `zxing_ReaderOptions` being opaque types is maybe not the most natural/canonical way to represent the API in c. I'm currently thinking about changing them to be simple structs. This would remove all the setter/getter/new/delete functions and allow them to be passed by value. I imagine a use case similar to this:
```c
    zxing_ReaderOptions opts { .tryHarder = true, .formats = zxing_BarcodeFormat_QRCode};
    opts.minLineCount = 1;

    zxing_ImageView iv { .data = NULL, .width = 100, .height = 20, .format = zxing_ImageFormat_RGB };

    zxing_Results* results = zxing_ReadBarcodes(iv, opts);
    zxing_Results_delete(results);
```

This would then even allow something as concise as the following:
```c
    results = zxing_ReadBarcodes((zxing_ImageView){NULL, zxing_ImageFormat_Lum, 100, 20}, zxing_ReaderOptions_default());
```
One thing that is still worrying me is the fact that bitfield structs like `ReaderOptions` are convenient in C but not so much when dealing with interfacing other languages (like Rust/.NET). So I would specify the C-version without this 'compression'. That means the `sizeof(zxing_ReaderOptions)` would be around 36 bytes (instead of 12). `sizeof(zxing_ImageView)` is 32.

Note that I recently already added a simple `struct zxing_Position` with direct field access.

@siiky / @hoisagen  do you see any issue with that approach (apart from being an API incompatible change...).

**Reply by siiky** [reply]
The opaque API has the advantage that it's harder to break (may not be so much of a problem in Rust; don't know about .NET), even if more verbose. You have more leeway in what you can do internally without breaking the external API (I'm probably preaching to choir here, you know better about this stuff than me for sure :).

If you want to keep the compactness of bitfields you might want to experiment with some bit fiddling. However, because `ReaderOptions` is not all bitfields I can only think of something like this, which is not very user-friendly.

```c
enum { TRY_HARDER_NO, TRY_HARDER_YES = 1<<0 };
enum { TRY_ROTATE_NO, TRY_ROTATE_YES = 1<<1 };
enum { TRY_INVERT_NO, TRY_INVERT_YES = 1<<2 };
/* ... */

(ReaderOptions){
  .bitfields=TRY_HARDER_YES|TRY_ROTATE_YES|TRY_INVERT_NO,
  .maxNumberOfSymbols=123,
  .formats=zxing_BarcodeFormat_QRCode,
  /* ... */
}
```

I just wonder if it's worth all the effort, assuming that .NET can call C procedures (Rust can do this for sure).

**Reply by axxel** [reply]
You are right about the increased ABI stability and yes, I certainly know enough about this stuff to be concerned. I don't see too much of a benefit in this case because the c-ABI version is tied to the c++ ABI version (`SONAME`) because they are part of the same binary. So if I change something on the c++ side and increase the `SONAME` that means the c-clients will need to be re-compiled anyway even if the c-ABI did not change at all. This is in and of itself not really a beneficial feature but I don't know how to prevent that other than distributing the c-API in a separate library, which I'd rather not because of the increased complexity.

If `ImageView` and `ReaderOptions` where anything like a mutable, stateful object (e.g. like the `struct sqlite3`) I would definitively stick to the opaque pointer. But they are actually just two collections of function parameters that are mainly there to allow to define a set of default values that don't have to be provided by the caller. In the Python API, the `ReaderOptions` struct does not exist because the language itself allows to selectively set optional parameters (like `read_barcodes(image, format=QRCode, tryHarder=true)`).

If the API would look like `void read_barcodes(char* data, int size, format_t format, bool return errors);` I would definitively not think about introducing a heap-allocated struct to pass the enum and the bool to the function via a pointer. And conceptually this is the same situation.

And don't worry, the manual bitfield manipulation via enum flags is not going to happen. That would be kind of the worst of both worlds.

@gitlost have you ever regretted that `struct zint_symbol` is a transparent struct? How do you deal with ABI breaking changes?

**Reply by siiky** [reply]
Actually I was thinking more from the perspective of the C API user/wrapper implementer. Recompiling is still less work than recompiling **and** having to fix code because the public API has changed.

If you go the transparent C struct route, any time you change the struct, you (likely) have also to change the wrappers to use the new fields or w.e.
If you keep the opaque C struct, you at most have to add/remove public procedures, everything else stays the same, even if you have to recompile code to use the new library version.

(btw I'm neutral here, just playing devil's advocate, both options have their merits)

**Reply by axxel** [reply]
Working on the .NET wrapper and its struct marshalling characteristics and @gitlost's comment about intermixing header files convinced me that keeping more boilerplate is indeed the better option (at least for now).

**Comment by gitlost** [comment]
Re `struct zint_symbol` transparent, yes, it does cause issues, e.g.
this recent ticket https://sourceforge.net/p/zint/tickets/303/.

Re dealing with ABI breaking changes, I don't! - just increment the version.

**Comment by axxel** [comment]
I just released an alpha version of the .NET wrapper an it turns out, I had bug in the handling of the return value of `zxing_ReadBarcodes()`: I thought a `NULL` ptr would mean there was an error executing the function, but a NULL value can mean 'empty list' or 'error'. Since I spent days thinking about this stuff and working with it, this was *really* embarrassing. I can only draw one conclusion from that experience: the decision to have NULL mean two things was ill advised. Before there was any error reporting that made sense because one did not have to check for NULL and call `zxing_Barcodes_size()` just to find out if there was any result or not. But now it is *way* more consistent and less error prone if NULL means error and everything else is a list of barcodes and that list might very well be empty. So much for an explanation why this behavior will change.

**Reply by axxel** [reply]
Related to this is the question of what to do with `zxing_ReadBarcode()` (singular)? It also returns `NULL` on both error and not-found conditions. It would also be possible to return a default constructed `Barcode`, where `zxing_Barcode_isValid()` returns false. But that I don't like at all. Having one function return an empty list but the other NULL instead of 'empty Barcode' is also making me gringe inside. I'm inclined to simply remove the singular version. On the c++-side it is there mostly because it has been there for ever / backward compatibility. The ios/Android/Rust/dotnet wrappers all don't have that version.

Is anyone really interested in keeping that function? Note: it is obviously possible to write a little helper function on the client side that provides the current functionality based on `zxing_ReadBarcodes()`.

**Reply by axxel** [reply]
Update: I removed the function today (before publicly releasing the C-API as part of the 2.3.0 release).

**Comment by axxel** [comment]
Another addition to the c-API is this `ImageView` constructor:
```c
zxing_ImageView* zxing_ImageView_new_checked(
    const uint8_t* data, int size, int width, int height, zxing_ImageFormat format, int rowStride, int pixStride);
```
which includes a new `size` parameter that allows to check for parameter consistency, i.e. out of bounds checks with the image dimension and layout stuff. This seemed very useful in the wrapper cases where there is always a block of memory with a size available.

I was contemplating whether this new function should be called `_new` and the old one `_new_unchecked` to kind of set a 'secure' default?

**Comment by pranavtriguna** [comment]
Following, since I am trying to evaluate if I can use this at work, preferably in C, but cpp may also work.
Thanks!

**Comment by axxel** [comment]
While summing up the history and the current state of the naming situation in #720, I realized that the prefix `zxing_` used in the C-API is rather unnecessarily inconsistent with the rest of the project. Based on that list, it might make more sense to use the prefix `zxingcpp_` or `ZXing_` or `ZXingCpp_`. Preferences?

**Reply by axxel** [reply]
Apart from the prefix, the rest of the naming is strictly following the C++ symbol names using `_` to replace wither `::` or `.`. So in that sense the prefix should be `ZXing_` to replace the `ZXing` c++ namespace. `ZXing::ReaderOptions::tryHarder` becomes `ZXing_ReaderOptions_tryHarder`. For lack of other suggestions, this is where I'll go then.

**Reply by siiky** [reply]
af I totally forgot to reply to this
I vote on either `zxing` or `ZXing` for being shorter; your reasoning about the `ZXing` makes sense, SGTM!

**Comment by SohayebHossain** [comment]
I cannot compile the wrapper C API for MCU (AC7911B, PI32V2 r3.

**Reply by axxel** [reply]
Thanks for letting us know but unfortunately, your comment does not contain any actionable information. If you'd like to get feedback that is supposed to help you fix your problem, then please open an issue and post the specific error message your get from your compiler. If I had to guess blindly, I'd say probably insufficient c++20 support.

**Reply by SohayebHossain** [reply]
In my project all code is in the C language. I want to use this JLAC791N MCU. So that I only used the C wrapper API. I already build it and created a binary file.
<img width="1521" height="423" alt="Screenshot 2026-03-16 085835" src="https://github.com/user-attachments/assets/c6e604dd-64aa-4a78-9fcb-422c3bcb44ce" />

**Reply by axxel** [reply]
Just googling for the error message showed me this: https://stackoverflow.com/questions/61646912/using-soil-lib-with-gcc-error-adding-symbols-file-format-not-recognised. It looks like you have an improperly setup cross compilation environment. I'm afraid I can't help you with that.

**Reply by SohayebHossain** [reply]
Do you have ZXing library with all are in C language and the binary .a file???

**Reply by axxel** [reply]
No.
